`E731` autofix broken for class vars

<!--
Thank you for taking the time to report an issue! We're glad to have you involved with Ruff.

If you're filing a bug report, please consider including the following information:

* A minimal code snippet that reproduces the bug.
* The command you invoked (e.g., `ruff /path/to/file.py --fix`), ideally including the `--isolated` flag.
* The current Ruff settings (any relevant sections from your `pyproject.toml`).
* The current Ruff version (`ruff --version`).
-->

(`E731` is "do not assign a `lambda` expression, use a `def`")

Consider this example where a `dataclass` has a `Callable` attribute with a default value:

```python
from collections.abc import Callable
from dataclasses import dataclass

@dataclass
class Foo:
    bar: Callable[[], str] = lambda: "foo"
```

Ruff identifies the class variable assignment as a violation of `E731`:

```console
$ ruff check test.py --isolated
test.py:7:5: E731 [*] Do not assign a `lambda` expression, use a `def`
Found 1 error.
[*] 1 potentially fixable with the --fix option.
```

But the autofix transforms it into:

```python
from dataclasses import dataclass

@dataclass
class Foo:
    def bar():
        return "foo"
```

which is not equivalent. 

Looking at https://github.com/charliermarsh/ruff/blob/9168a12679976ece5e94dd9cf6abe3c8886e31cc/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs#L29-L37 it seems like the problem is that there's no check on whether it makes sense to do this transformation in the surrounding context (in this case, the autofix doesn't apply because we're in a `@dataclass` body, but maybe there are other contexts too?). 

I think the desired autofix in this case would be more like:

```python
from collections.abc import Callable
from dataclasses import dataclass

def bar():
    return "foo"

@dataclass
class Foo:
    bar: Callable[[], str] = bar
```

but that is only correct for dataclasses because we care about the type-annotated class variable! If this wasn't a dataclass, the original autofix would be fine and produces equivalent behavior at runtime. I'd suggest that as a first pass, there's just no autofix for `E731` if it occurs in a class body, because it depends on this question of whether the type-annotated class variable needs to exist for some other code to use, which is probably impossible to determine (since it's not just `@dataclass` that might use those annotations).

Happy to take a look at a fix for this issue myself, but I'm a first-time contributor so it might take some time :)

--- a/crates/ruff/resources/test/fixtures/pycodestyle/E731.py
+++ b/crates/ruff/resources/test/fixtures/pycodestyle/E731.py
@@ -9,6 +9,9 @@ while False:
 f = lambda: (yield 1)
 #: E731
 f = lambda: (yield from g())
+#: E731
+class F:
+    f = lambda x: 2 * x
 
 f = object()
 f.method = lambda: "Method"

--- a/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
+++ b/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
@@ -25,12 +25,8 @@ impl Violation for LambdaAssignment {
     }
 
     fn autofix_title_formatter(&self) -> Option<fn(&Self) -> String> {
-        let LambdaAssignment { fixable, .. } = self;
-        if *fixable {
-            Some(|v| format!("Rewrite `{}` as a `def`", v.name))
-        } else {
-            None
-        }
+        self.fixable
+            .then_some(|v| format!("Rewrite `{}` as a `def`", v.name))
     }
 }
 

--- a/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
+++ b/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
@@ -8,20 +8,29 @@ use crate::checkers::ast::Checker;
 use crate::fix::Fix;
 use crate::registry::Diagnostic;
 use crate::source_code::Stylist;
-use crate::violation::AlwaysAutofixableViolation;
+use crate::violation::{AutofixKind, Availability, Violation};
 
 define_violation!(
-    pub struct LambdaAssignment(pub String);
+    pub struct LambdaAssignment {
+        pub name: String,
+        pub fixable: bool,
+    }
 );
-impl AlwaysAutofixableViolation for LambdaAssignment {
+impl Violation for LambdaAssignment {
+    const AUTOFIX: Option<AutofixKind> = Some(AutofixKind::new(Availability::Sometimes));
+
     #[derive_message_formats]
     fn message(&self) -> String {
         format!("Do not assign a `lambda` expression, use a `def`")
     }
 
-    fn autofix_title(&self) -> String {
-        let LambdaAssignment(name) = self;
-        format!("Rewrite `{name}` as a `def`")
+    fn autofix_title_formatter(&self) -> Option<fn(&Self) -> String> {
+        let LambdaAssignment { fixable, .. } = self;
+        if *fixable {
+            Some(|v| format!("Rewrite `{}` as a `def`", v.name))
+        } else {
+            None
+        }
     }
 }
 
@@ -29,8 +38,13 @@ impl AlwaysAutofixableViolation for LambdaAssignment {
 pub fn lambda_assignment(checker: &mut Checker, target: &Expr, value: &Expr, stmt: &Stmt) {
     if let ExprKind::Name { id, .. } = &target.node {
         if let ExprKind::Lambda { args, body } = &value.node {
-            let mut diagnostic =
-                Diagnostic::new(LambdaAssignment(id.to_string()), Range::from_located(stmt));
+            let mut diagnostic = Diagnostic::new(
+                LambdaAssignment {
+                    name: id.to_string(),
+                    fixable: true,
+                },
+                Range::from_located(stmt),
+            );
             if checker.patch(diagnostic.kind.rule()) {
                 if !match_leading_content(stmt, checker.locator)
                     && !match_trailing_content(stmt, checker.locator)

--- a/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
+++ b/crates/ruff/src/rules/pycodestyle/rules/lambda_assignment.rs
@@ -2,7 +2,7 @@ use ruff_macros::{define_violation, derive_message_formats};
 use rustpython_parser::ast::{Arguments, Expr, ExprKind, Location, Stmt, StmtKind};
 
 use crate::ast::helpers::{match_leading_content, match_trailing_content, unparse_stmt};
-use crate::ast::types::Range;
+use crate::ast::types::{Range, ScopeKind};
 use crate::ast::whitespace::leading_space;
 use crate::checkers::ast::Checker;
 use crate::fix::Fix;
@@ -38,42 +38,52 @@ impl Violation for LambdaAssignment {
 pub fn lambda_assignment(checker: &mut Checker, target: &Expr, value: &Expr, stmt: &Stmt) {
     if let ExprKind::Name { id, .. } = &target.node {
         if let ExprKind::Lambda { args, body } = &value.node {
+            // If the assignment is in a class body, it might not be safe
+            // to replace it because the assignment might be
+            // carrying a type annotation that will be used by some
+            // package like dataclasses, which wouldn't consider the
+            // rewritten function definition to be equivalent.
+            // See https://github.com/charliermarsh/ruff/issues/3046
+            let fixable = !matches!(checker.current_scope().kind, ScopeKind::Class(_));
+
             let mut diagnostic = Diagnostic::new(
                 LambdaAssignment {
                     name: id.to_string(),
-                    fixable: true,
+                    fixable,
                 },
                 Range::from_located(stmt),
             );
-            if checker.patch(diagnostic.kind.rule()) {
-                if !match_leading_content(stmt, checker.locator)
-                    && !match_trailing_content(stmt, checker.locator)
+
+            if checker.patch(diagnostic.kind.rule())
+                && fixable
+                && !match_leading_content(stmt, checker.locator)
+                && !match_trailing_content(stmt, checker.locator)
+            {
+                let first_line = checker.locator.slice(&Range::new(
+                    Location::new(stmt.location.row(), 0),
+                    Location::new(stmt.location.row() + 1, 0),
+                ));
+                let indentation = &leading_space(first_line);
+                let mut indented = String::new();
+                for (idx, line) in function(id, args, body, checker.stylist)
+                    .lines()
+                    .enumerate()
                 {
-                    let first_line = checker.locator.slice(&Range::new(
-                        Location::new(stmt.location.row(), 0),
-                        Location::new(stmt.location.row() + 1, 0),
-                    ));
-                    let indentation = &leading_space(first_line);
-                    let mut indented = String::new();
-                    for (idx, line) in function(id, args, body, checker.stylist)
-                        .lines()
-                        .enumerate()
-                    {
-                        if idx == 0 {
-                            indented.push_str(line);
-                        } else {
-                            indented.push_str(checker.stylist.line_ending().as_str());
-                            indented.push_str(indentation);
-                            indented.push_str(line);
-                        }
+                    if idx == 0 {
+                        indented.push_str(line);
+                    } else {
+                        indented.push_str(checker.stylist.line_ending().as_str());
+                        indented.push_str(indentation);
+                        indented.push_str(line);
                     }
-                    diagnostic.amend(Fix::replacement(
-                        indented,
-                        stmt.location,
-                        stmt.end_location.unwrap(),
-                    ));
                 }
+                diagnostic.amend(Fix::replacement(
+                    indented,
+                    stmt.location,
+                    stmt.end_location.unwrap(),
+                ));
             }
+
             checker.diagnostics.push(diagnostic);
         }
     }

